Sage/carbon
# fix(select): ensure aria-activedescendant is always valid

make the Select List always part of the DOM, and hide it with CSS instead. This allows the component
to ensure that the guids provided as IDs to the Option children are always stable, so that the
aria-activedescendant attribute of the Select Textbox always refers to the ID of an existing
element. Also make the aria-controls attribute of the textbox always present so that screenreaders
can know the list of options available when focused on the textbox.

fix #5289

### Proposed behaviour

The Select textbox should always have an aria-controls attribute that refers to an option list that is always in the DOM, and is hidden and shown with CSS - similar to the ARIA Working Practices example here https://www.w3.org/WAI/ARIA/apg/example-index/combobox/combobox-select-only.html

Further, the IDs of each Option element should be stable (assuming no changes to the available options), so that the aria-activedescendant attribute, which changes as focus is moved among options with the keyboard as well as when selection is made with the mouse, always holds the actual ID of an element that is in the DOM (even if hidden).

There has been a further small change to the FilterableSelect component - this was purely to fix a failing cypress test which seemingly only happened with this PR and proved to be due to a race condition when handleGlobalClick runs at the same time as an update to the `value` prop in the controlled case.



### Current behaviour

The Select option list is removed from the DOM when closed, so that when selecting an option and closing the list, the aria-activedescendant is invalid, causing Axe to flag an error.

Further, even when reopening the select so that the options are in the DOM again, the IDs are newly randomly-generated guids so the existing aria-activedescendant is still not referring to an element that's present in the DOM.



### Checklist




#### QA


### Additional context



### Testing instructions


Test Stories for each type of Select (Simple select, FilterableSelect, MultiSelect). Select an option, close the select, and run Axe devtools. Reopen the select (without choosing a new option) and run again. Ensure that no errors are flagged (or at least not one saying ARIA attribute values are invalid!)

Also test with screenreaders across all supported screenreader/browser combinations, to ensure the Selects are usable - or at least there are no issues introduced here which didn't exist before.

Regression testing should also be done with FilterableSelect to ensure that the change I've made to the controlled case (see under "proposed behaviour" above, final paragraph) does not cause any unforeseen issues.
